File: perl/hop/4/Permutations.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class Permutations {
	
	static <E> void Permute(List<E> items, List<E> perms) {
		if ( items.isEmpty() ) {
			Print(perms);
		}else {
			for ( int i=0; i<items.size(); i++ ) {
				List<E> newItems = new LinkedList<E>(items);
				List<E> newPerms = new LinkedList<E>(perms);
				newPerms.add(0, newItems.get(i));
				newItems.remove(i);
				Permute(newItems, newPerms);
			}
		}
    }
	
	static <E> void Print(List<E> items) {
		for (E e : items) {
			System.out.print(e+" ");
		}
		System.out.println();
	}
	
	public static void main(String args[]) {
		List<String> list = new ArrayList<String>(
					Arrays.asList("1", "2", "3", "4"));
		Permute(list, new LinkedList<String>());
	}
}

File: perl/hop/4/permutations.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub permute {
	my @items = @{ $_[0] };
	my @perms = @{ $_[1] };
	unless (@items) {
		print "@perms\n";
	} else {
		my ( @newitems, @newperms, $i );
		foreach $i ( 0 .. $#items ){
			@newitems = @items;
			@newperms = @perms;
			unshift( @newperms, splice( @newitems, $i, 1 ) );
			permute( [@newitems], [@newperms] );
		}
	}
}

permute( [qw/1 2 3 4 /], [] );
